### DISCUSSION TITLE
调度系统设计精要 · /system-design-scheduler

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#589] [Announcements]
https://draveness.me/system-design-scheduler

**Comment by yuchanns** [comment]
今天刚跟朋友聊起`scheduler`，draveness就发了篇调度文章，nice!

**Comment by draveness** [comment]
> @yuchanns
> 今天刚跟朋友聊起`scheduler`，draveness就发了篇调度文章，nice!

希望有点用哈

**Comment by cherrot** [comment]
提醒一下，twitter分享生成的链接不对 :)

**Comment by draveness** [comment]
> 提醒一下，twitter分享生成的链接不对 :)

咦，我看没有问题啊，是 Tweet 里的链接么？

**Comment by cherrot** [comment]
> @draveness
> > 提醒一下，twitter分享生成的链接不对 :)
> 
> 咦，我看没有问题啊，是 Tweet 里的链接么？

域名和URI之间少了斜线/，生成的是:
> 调度系统设计精要 https://draveness.mesystem-design-scheduler

**Comment by draveness** [comment]
> @cherrot
> > @draveness
> > > 提醒一下，twitter分享生成的链接不对 :)
> > 
> > 咦，我看没有问题啊，是 Tweet 里的链接么？
> 
> 域名和URI之间少了斜线/，生成的是:
> > 调度系统设计精要 https://draveness.mesystem-design-scheduler

看到了，多谢

**Comment by renjingneng** [comment]
我擦 这篇文章无敌了 
概念理解起来不难
但是要是深入到源码  有相当的理解真的很难 慢慢啃

**Comment by draveness** [comment]
> 我擦 这篇文章无敌了
> 概念理解起来不难
> 但是要是深入到源码 有相当的理解真的很难 慢慢啃

😂😂😂

**Comment by Reasno** [comment]
太厉害了

**Comment by schwarzeni** [comment]
厉害 👍👍👍

**Comment by daoshenzzg** [comment]
👍

**Comment by Icedroid** [comment]
演讲中提到kubernetes的调度框架可以直接引用在我们的业务上，这个有具体的demo样例吗，想了解下如果拆分出类似的调度设计在自己的业务上使用。

**Comment by draveness** [comment]
> 演讲中提到kubernetes的调度框架可以直接引用在我们的业务上，这个有具体的demo样例吗，想了解下如果拆分出类似的调度设计在自己的业务上使用。

这个没有，不过我可以搞一个

**Comment by FingerLiu** [comment]
[typo]
Kubernetes 迅速占领市场，成为容器编排领域的实施标准。 -> Kubernetes 迅速占领市场，成为容器编排领域的事实标准。

**Comment by draveness** [comment]
> [typo]
> Kubernetes 迅速占领市场，成为容器编排领域的实施标准。 -> Kubernetes 迅速占领市场，成为容器编排领域的事实标准。

我改一下

**Comment by Icedroid** [comment]
> > 演讲中提到kubernetes的调度框架可以直接引用在我们的业务上，这个有具体的demo样例吗，想了解下如果拆分出类似的调度设计在自己的业务上使用。
> 
> 这个没有，不过我可以搞一个

期待

**Comment by fwensen** [comment]
在我们的业务场景上有个调度问题，一批任务生成到数据库中，然后由另外的机器从中拉取任务执行，这些任务有耗时长，耗时短的，同时有优先级(同一类型任务有来源方的优先级)，不同任务底层是调用另外的服务，即不同任务有不同的qps，我们希望尽量每个任务都能达到qps极限，同时又不影响优先级，博主有什么建议吗

**Comment by draveness** [comment]
> 在我们的业务场景上有个调度问题，一批任务生成到数据库中，然后由另外的机器从中拉取任务执行，这些任务有耗时长，耗时短的，同时有优先级(同一类型任务有来源方的优先级)，不同任务底层是调用另外的服务，即不同任务有不同的qps，我们希望尽量每个任务都能达到qps极限，同时又不影响优先级，博主有什么建议吗

如果仅仅想要达到 QPS 极限，就先按优先级排，再按执行时间倒排，由于提供的上下文有限，这个策略假设了优先级是必须满足的硬限制。

如果想要设计更完备的系统，可能存在下面的问题需要思考：

1. 是不是需要考虑可能存在的饥饿问题，耗时长和低优先级的任务迟迟得不到执行，如果不考虑这个问题，就很好办了；
2. 哪些条件在你的调度系统里是必须满足的硬限制，哪些是尽量满足的软限制，例如：优先级是必须满足的硬限制，任务的到期时间可能是软限制；

**Comment by yanjinbin** [comment]
可以热心翻译一波英文？
精品。
不过更喜欢楼主如何挖掘这些文章的过程，以及如何挖掘的技术和思路。

**Comment by draveness** [comment]
> 可以热心翻译一波英文？
> 精品。
> 不过更喜欢楼主如何挖掘这些文章的过程，以及如何挖掘的技术和思路。

还有这个操作呢...

**Comment by yanjinbin** [comment]
> > 可以热心翻译一波英文？
> > 精品。
> > 不过更喜欢楼主如何挖掘这些文章的过程，以及如何挖掘的技术和思路。
> 
> 还有这个操作呢...

额。。。说的不太对。。我只是想说如何写出这篇好文章，中间经历的搜集材料以及验证，用到的技术和思路的过程

**Comment by draveness** [comment]
> > > 可以热心翻译一波英文？
> > > 精品。
> > > 不过更喜欢楼主如何挖掘这些文章的过程，以及如何挖掘的技术和思路。
> > 
> > 
> > 还有这个操作呢...
> 
> 额。。。说的不太对。。我只是想说如何写出这篇好文章，中间经历的搜集材料以及验证，用到的技术和思路的过程

有这个打算，不过你说翻译英文是要干啥

**Comment by yanjinbin** [comment]
> > > > 可以热心翻译一波英文？
> > > > 精品。
> > > > 不过更喜欢楼主如何挖掘这些文章的过程，以及如何挖掘的技术和思路。
> > > 
> > > 
> > > 还有这个操作呢...
> > 
> > 
> > 额。。。说的不太对。。我只是想说如何写出这篇好文章，中间经历的搜集材料以及验证，用到的技术和思路的过程
> 
> 有这个打算，不过你说翻译英文是要干啥

当时的意思是说，很难见到 『从全局和历史发展脉络展开，又或者是 深入底成分析调用链的文章』,
想必这写成这样的技术文章，需要比较很强的专业知识储备。
这样好的文章，除了中文技术圈，还可以向英文技术圈传播。
反正，我觉得写的比 ardan 那个人好。
朴素而又硬核的文章，要珍惜。

即便是英文的go技术文章，我还是比较难以碰到好文啊，更多的传授比较浅显的技巧或者经验

所以，我更好奇的是楼主本身
 1 前期的技术知识储备是如何积累的？ 平平无奇的技术博客，靠着copy和拼凑出来的文章，大多没有什么价值
2 有了这知识储备之后，技术写作 写得好，需要从一个恰当的点切入展开叙述，抓住核心知识点或者代码块，这中间过程可能反反复复修改。
3 u1s1, 很admire 楼主的知识储备和写作能力。😄，请问，怎么样做到的？愿意付费喔

**Comment by draveness** [comment]
> 当时的意思是说，很难见到 『从全局和历史发展脉络展开，又或者是 深入底成分析调用链的文章』,
> 想必这写成这样的技术文章，需要比较很强的专业知识储备。
> 这样好的文章，除了中文技术圈，还可以向英文技术圈传播。
> 反正，我觉得写的比 ardan 那个人好。
> 朴素而又硬核的文章，要珍惜。
> 
> 即便是英文的go技术文章，我还是比较难以碰到好文啊，更多的传授比较浅显的技巧或者经验

感谢支持，不过我不知道 ardan 是谁

> 所以，我更好奇的是楼主本身
> 1 前期的技术知识储备是如何积累的？ 平平无奇的技术博客，靠着copy和拼凑出来的文章，大多没有什么价值

看论文、看书

> 2 有了这知识储备之后，技术写作 写得好，需要从一个恰当的点切入展开叙述，抓住核心知识点或者代码块，这中间过程可能反反复复修改。

确实修改了很多次，删了很多章节

> 3 u1s1, 很admire 楼主的知识储备和写作能力。😄，请问，怎么样做到的？愿意付费喔

我不一定能给出多**有用并且可行**的建议，可以 Q & A 的形式聊聊，公众号里有我的微信

**Comment by edisonwd** [comment]
**typo:**

**原文：充分利用多和 CPU 的并行能力**

**修改：充分利用多核 CPU 的并行能力**

---

*2020-10-27 UPDATES: 已修复*

**Comment by ppd0705** [comment]
之前看过一遍，今天看完《Operating Systems: Three Easy Pieces》相关部分再来看一遍理解更深了一点点

**Comment by existentialist-pc** [comment]
Typo:


原文：完成跨度（Makesapan）

修改：完成跨度（Makespan）

**Comment by changyonggang** [comment]
高山啊！

**Comment by wxuedong** [comment]
公众号的Weixin ID 找不到用户

**Comment by draveness** [comment]
> 公众号的Weixin ID 找不到用户

我可以扫出来啊

**Comment by wxuedong** [comment]
> @draveness
> > 公众号的Weixin ID 找不到用户
> 
> 我可以扫出来啊

不是扫一扫，是通过Weixin 添加搜索用户

**Comment by draveness** [comment]
> > @draveness
> > > 公众号的Weixin ID 找不到用户
> > 
> > 
> > 我可以扫出来啊
> 
> 不是扫一扫，是通过Weixin 添加搜索用户

搜索『真没什么逻辑』就行了

**Comment by wxuedong** [comment]
老师 你好，我想在descheduler 扩展一个基于Node负载（load average）增高进行调度策略，比如 一个Node负载增高到设置负载阀值 开始驱逐当前Node 上的Pods，现在遇到一个问题就是如何定位到在驱逐的这些Pods中识别并不是制造Node 增高的元凶，或者有没有其他思路避免把制造高负载的Pod 驱逐到其他Node上 ？

**Comment by draveness** [comment]
> 老师 你好，我想在descheduler 扩展一个基于Node负载（load average）增高进行调度策略，比如 一个Node负载增高到设置负载阀值 开始驱逐当前Node 上的Pods，现在遇到一个问题就是如何定位到在驱逐的这些Pods中识别并不是制造Node 增高的元凶，或者有没有其他思路避免把制造高负载的Pod 驱逐到其他Node上 ？

你们的 Pod 不是 Guaranteed 的么，如果不是的话就监控每个 Pod 的 CPU 时间呗，哪个有 Spike 触发驱逐的时候就把它迁移走，不过这样会影响程序的 QoS 的，业务不会投诉么

**Comment by unknwon** [comment]
捉个虫 😆
> Kubernetes 从 v1.0.0 版本发布到 v1.14.0，总共 15 个版本一直都在使用谓词和优先级来管理不同的调度算法，知道 v1.15.0 开始引入调度框架

s/知道/直到/

---

*2020-02-28 UPDATES：已修复*

**Comment by kaydxh** [comment]
这个网站、专属域名怎么搭建的哈？

**Comment by heng4fun** [comment]
系统好文, 看完后对于调度有了更加整体的了解.

- 一些扩展资料的链接不对了, 比如 K8s 的 scheduling-framework 的 KEP 链接, 应该是他们换了仓库地址; 
- IBM 的 <Create a custom Kubernetes scheduler> 地址可以换成下面这个新地址
https://developer.ibm.com/articles/creating-a-custom-kube-scheduler/

**Comment by WBFF95** [comment]
看了您的文章，受益匪浅。我最近也有在做一些关于调度的工作，有很多想请教的地方，怎么跟您联系呢？
